### PR TITLE
regression: add tests for the supplicant plugin framework

### PR DESCRIPTION
**Pull request by anisyanka**
Add test cases for checking functionalitu of supplicant plugins:
 - Ability to pass values and arrays to/from a plugin
 - Right handle situations with bad input parameters
 - Right handle situations with calling an unknown plugin

Signed-off-by: Aleksandr Anisimov <a.anisimov@omprussia.ru>

### CODE REVIEW

**Comment by anisyanka** [comment]
Relates to:
https://github.com/OP-TEE/optee_client/pull/239
https://github.com/OP-TEE/optee_os/pull/4248
https://github.com/linaro-swg/optee_examples/pull/79

File: host/supp_plugin/test_supp_plugin.c [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+		return get_test_arr(sub_cmd, data, data_len, out_len);
+	default:
+		break;
+	}
+
+	return TEEC_ERROR_NOT_SUPPORTED;
+}
+
+struct plugin_method plugin_method = {
+	"test",
```
Comment (jenswi-linaro) [inline host/supp_plugin/test_supp_plugin.c:102]: Please name the fields you're initializing. The NULL below can be skippen then also.

Comment (anisyanka) [reply host/supp_plugin/test_supp_plugin.c]: Done

**Comment by anisyanka** [comment]
All comments have been addressed

File: host/xtest/regression_1000.c [hunk 2]
```diff
@@ -2345,3 +2349,142 @@ static void xtest_tee_test_1032(ADBG_Case_t *c)
...
+
+		op.params[0].value.a = 20;
+		op.params[0].value.b = 10;
+		op.params[1].value.a = '+';
+		op.paramTypes = TEEC_PARAM_TYPES(TEEC_VALUE_INOUT,
+						 TEEC_VALUE_INPUT, TEEC_NONE,
+						 TEEC_NONE);
+
+		ADBG_EXPECT_TEEC_SUCCESS(c, TEEC_InvokeCommand(&session,
+					 TA_SUPP_PLUGIN_PASS_VALUES, &op,
```
Comment (jenswi-linaro) [inline host/xtest/regression_1000.c:2383]: You're aligning with wrong `(`
This comment applies to a few places below too.

File: ta/supp_plugin/include/ta_supp_plugin_test.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ * Copyright (c) 2021 Open Mobile Platform LLC
+ */
+
+#ifndef TA_SUPP_PLUGIN_TEST_H
+#define TA_SUPP_PLUGIN_TEST_H
+
+#define TA_SUPP_PLUGIN_TEST_UUID { 0x380231ac, 0xfb99, 0x47ad, \
+	{ 0xa6, 0x89, 0x9e, 0x01, 0x7e, 0xb6, 0xe7, 0x8a } }
+
+#define TA_SUPP_PLUGIN_PING		0
```
Comment (jenswi-linaro) [inline ta/supp_plugin/include/ta_supp_plugin_test.h:12]: Please rename these TA commands as `TA_SUPP_PLUGIN_CMD_*`

File: ta/supp_plugin/include/ta_supp_plugin_test.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2021 Open Mobile Platform LLC
+ */
+
+#ifndef TA_SUPP_PLUGIN_TEST_H
+#define TA_SUPP_PLUGIN_TEST_H
+
+#define TA_SUPP_PLUGIN_TEST_UUID { 0x380231ac, 0xfb99, 0x47ad, \
```
Comment (jenswi-linaro) [inline ta/supp_plugin/include/ta_supp_plugin_test.h:9]: Please rename to `TA_SUPP_PLUGIN_UUID`, the `TEST` part is redundant.
Please drop the `_test` part from the name of this file too.

File: host/xtest/regression_1000.c [hunk 5]
```diff
@@ -2345,3 +2349,142 @@ static void xtest_tee_test_1032(ADBG_Case_t *c)
...
+
+static void xtest_tee_test_1033(ADBG_Case_t *c)
+{
+	TEEC_Result res = TEEC_ERROR_GENERIC;
+	TEEC_Session session = { };
+	uint32_t ret_orig = 0;
+
+	/*
+	 * TA will ping the test plugin during open session operation. 
+	 * Supplicant plugins are optional: warn and nicely exit if not found
```
Comment (jenswi-linaro) [inline host/xtest/regression_1000.c:2361]: Optional? I can't see that in https://github.com/OP-TEE/optee_client/pull/239

File: ta/supp_plugin/ta_entry.c [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+				      __unused)
+{
+	const TEE_UUID nulluuid = {  }; /* null-uuid */
+	char data[32] = {  };
+	size_t inlen = sizeof(data);
+
+	return tee_invoke_supp_plugin(&nulluuid, 0, 0, data, inlen, NULL);
+}
+
+TEE_Result TA_InvokeCommandEntryPoint(void *pSessionContext __unused,
```
Comment (jenswi-linaro) [inline ta/supp_plugin/ta_entry.c:157]: All these TA functions could be replaced by a single TA function (the os_test TA would be a good match then) with a flexible interface allowing to put most of the logic in xtest. If you'd rather keep it as is that's fine too.

File: host/xtest/regression_1000.c (hunk 5 above)
Comment (anisyanka) [reply host/xtest/regression_1000.c:2361]: Fixed, thanks.

File: ta/supp_plugin/ta_entry.c (hunk 6 above)
Comment (anisyanka) [reply ta/supp_plugin/ta_entry.c:157]: I'd prefer to keep it as it is, thanks

**Comment by anisyanka** [comment]
All comments have been addressed

**Comment by jenswi-linaro** [comment]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by anisyanka** [comment]
Squashed the fixup commits ans update the commit message.
Thank you for the review!

**Comment by jforissier** [comment]
IBART is passing, all good! Thanks @anisyanka.

**Pull request merged**